docker/main/ngen/py_funcs.py

```diff
@@ -0,0 +1,309 @@
+    logging.basicConfig(
+        level=logging.getLevelName(log_level),
+        format="%(asctime)s,%(msecs)d %(levelname)s: %(message)s",
+        datefmt="%H:%M:%S"
+    )
+
+
+def _parse_for_tar_and_copy(parent_subparsers_container):
+    """ Run subparser for CLI command responsible for running `tar_and_copy` helper function. """
+    # A parser for the 'tar_and_copy' param itself, underneath the parent 'command' subparsers container
+    helper_cmd_parser = parent_subparsers_container.add_parser(
+        'tar_and_copy', description="Archive contents of a directory and copy archive to specified destination.")
+    helper_cmd_parser.add_argument('--dry-run', dest='do_dry_run', action='store_true',
+                                   help='Perform a dry run to check paths, with no archiving/moving/copying.')
+    helper_cmd_parser.add_argument('--compress', dest='do_compress', action='store_true',
+                                   help='Compress the created archive with gzip compression.')
+    helper_cmd_parser.add_argument('archive_name', help='Basename for the created archive file.')
+    helper_cmd_parser.add_argument('source', type=Path, help='Directory whose contents should be archived.')
+    helper_cmd_parser.add_argument('dest', type=Path, help='Destination directory in which to place the archive file.')
+
+
+def _parse_for_gather_output(parent_subparsers_container):
+    # A parser for the 'tar_and_copy' param itself, underneath the parent 'command' subparsers container
+    desc = "Using subprocesses, gather output from remote MPI hosts and collect in analogous directory on this host."
+    helper_cmd_parser = parent_subparsers_container.add_parser('gather_output', description=desc)
+    # TODO: (later) when we move to all Python, rework this to accept several individually
+    helper_cmd_parser.add_argument('mpi_hosts_str', help='Comma delimited MPI hosts string received by worker.')
+    helper_cmd_parser.add_argument('output_write_dir', type=Path, help='Directory where output was written on hosts.')
+
+
+def _subparse_move_to_directory(parent_subparser_container):
+    sub_cmd_parser = parent_subparser_container.add_parser('to_directory', description="Move to a specified directory")
+    sub_cmd_parser.add_argument("dest_dir", type=Path, help="Destination directory to which to move the output")
+
+
+def _parse_for_move_job_output(parent_subparsers_container):
+    # A parser for the 'tar_and_copy' param itself, underneath the parent 'command' subparsers container
+    desc = "Move output data files produced by a job to another location, typically to put them into a DMOD dataset."
+    helper_cmd_parser = parent_subparsers_container.add_parser('move_job_output', description=desc)
+    helper_cmd_parser.add_argument('--job_id', '--job-id', dest='job_id', help='Optionally specify job id.')
+    helper_cmd_parser.add_argument('--archive-files', dest='do_archiving', choices=["true", "false"],
```

Comment:
Set boolean values via `action="store_true"`, not via choices and a function in the type.

Reply:
on the other hand, if this is supposed to be called programmatically via a bash function, set up a script to call this programmatically via python, _not_ by passing raw boolean values through the CLI.

Reply:
So the reason this was done like this is to have a ``True``: _force-on_, ``False``:_force-off_, ``None``:_dynamically-determine_ setup.  

I wanted this to be able to receive ``None`` downstream where the ``do_archiving`` arg value gets used, as that's when there's some code used to decide whether or not to archive based on the files themselves.   But I also wanted an explicit indication, either way, to be respected.  And using ``store_true`` can't get us both ``None`` and ``False`` values.

Whether or not this is a good idea is certainly open for discussion, but that was why this argument wasn't set up in a more typical manner.

Reply:
I _think_ I know a good way of doing this, but I need to poke around at it first.

Reply:
Code like the following should work _and_ provide a more elegant cli:

```python
    archive_group = helper_cmd_parser.add_mutually_exclusive_group()
    archive_group.add_argument(
        "--force-archives",
        dest="force_archives",
        action="store_true",
        help="Whatever the force on logic is"
    )
    archive_group.add_argument(
        "--prevent-archives",
        dest="prevent_archives",
        action="store_true",
        help="Whatever the force off logic is"
    )
```

You'd then modify:

```python
def move_job_output(output_directory: Path, move_action: str, do_archiving: Optional[bool] = None,
                    job_id: Optional[str] = None, **kwargs):
```

to be

```python
def move_job_output(output_directory: Path, move_action: str, force_archiving: bool = False,
                    prevent_archives: bool = False, job_id: Optional[str] = None, **kwargs):
```

which will have logic like:

```python
    # If this was not set, dynamically determine what it should be
    if not force_archiving and not prevent_archives:
        # For now, just do this if the output data contains more than 100 individual files
        out_dir_files = [f for f in output_directory.glob("**/*")]
        out_dir_file_count = len(out_dir_files)
```

If you want to make things a tad more explicit, you can use an enum:

```python
class ArchiveStrategy(enum.Enum):
    FORCE_ON = "force"
    FORCE_OFF = "prevent"
    DYNAMIC = "dynamic"

...

    helper_cmd_parser.add_argument(
        "--archiving",
        dest="archiving",
        default=ArchiveStrategy.DYNAMIC,
        type=ArchiveStrategy,
        help="Limited value with an enum"
    )

...

def move_job_output(output_directory: Path, move_action: str, archiving: ArchiveStrategy = ArchiveStrategy.DYNAMIC,
    job_id: Optional[str] = None, **kwargs):

...
   if archiving == ArchivingStrategy.DYNAMIC:
        # For now, just do this if the output data contains more than 100 individual files
        out_dir_files = [f for f in output_directory.glob("**/*")]

```

Reply:
I don't actually think this is better here.  For one, I don't particularly like that this leads to the possibility, and thus the need to handle the condition, when:

`force_archiving == True and prevent_archives == True`

I'd rather structure things to avoid the possibility of the contradiction altogether.

Beyond that, we also can assume we'll be changing how the CLI parsing and behavior works when we do #665.  Aside from this upcoming flux complicating a plan to prevent the aforementioned contradiction at the CLI level, I'd rather prioritize the design of the underlying function in this case.

I'm also a little on the fence as to whether `--force-this` / `--disallow-this` is generally easier to use and maintain than `--do-this [true|false]`.  I could certainly be convinced there, but I think that's another thing that should wait until #665.

Reply:
I lurch away from `--do-this [true|false]` because it's a soft enum and/or treating a boolean as a qbit. It's not incorrect or non-functional per se, but it's still something that I'd consider bad practice. If you're going to deal with boolean represented values in a CLI, that's generally relegated to the world of flags and a set of constants would generally be a set of constants even when the three states of a boolean may be used to represent the constant.

For your bit about `force_archiving == True and prevent_archives == True` - the mutually exclusive group will prevent that case from ever occurring - the `ArgumentParser` will throw an error before you ever get there.

I'll leave it to Austin to decide if I'm being a little too hardlined on this detail. I don't want an opinion about an interface philosophy to derail anyone's work.